Workshop Class, Session Class, and Student Class

Chris-Pyles opened the pull request:
adds workshop class used to describe each workshop and all its properties including
+ name
+ number
+ capacity (total and per session)
+ popularity score
+ variables to check if a session is full or "full enough"
+ function to update all aforementioned variables
+ function to add student to a given session
+ function to set a session's capacity manually

NOTE: I am aware of the TODO still on line 12. Since this number will be calculated pretty subjectively, I feel that it's better to have the class and its framework set up in the master branch for further programming, since this number will probably be changed a lot anyway as we write the actual algorithm.

Chris-Pyles committed:
Added Workshop "class"

Workshop isn't really a class (we can't actually use `class` in google apps script) but it functions exactly as one.

Chris-Pyles committed:
added function to workshop class

new function setSessionCapacity sets the capacity to a given value for a certain session (A, B, or C) in the workshop.

Chris-Pyles committed:
Added TODOs

Added TODOs dependent on other work from other branches
-line 12
-line 67
-line 68

Chris-Pyles committed:
bug fixes

fixed error on line 102 (changed isFullB to isFullC)

Chris-Pyles committed:
Improvements to workshop class

+ added global POPULARITY_POINTS variable, an array used to calculate the points for a workshop's popularity score based on it appearing as a girl's 1st, 2nd, etc choice in their preferences
+ workshop popularity score is now calculated automatically when the workshop object is constructed
+ added makeWorkshopArray function that constructs an array of all workshops in the workshop sheet
+ added global WORKSHOP_ARRAY variable, an array of all workshops sorted by popularity
+ added documentation for workshop class and all its class functions
- popularityRank class variable has been removed from the workshop class
- removed workshopTesting function

Chris-Pyles committed:
adjusted setWorkshopCapacity function behavior

setWorkshopCapacity method of the workshop class now also changes the workshop's static capacity variables

beszel commented on main.js:
```diff
@@ -33,6 +41,171 @@ function onOpen() {
+                this.popularityScore += POPULARITY_POINTS[j];
+            }
+        }
+    }
```
Is it appropriate to refactor this into a method? I think it would help simplify the constructor.

beszel commented on main.js:
```diff
@@ -33,6 +41,171 @@ function onOpen() {
+    /**
+     * Recalculates important class variables used for analysis based on changes to other ones
+     */
+    this.statusUpdate = function() {
```
The boolean properties could be replaced with methods that perform these calculations when called.

Chris-Pyles committed:
calculating popularity score is now a method that is called in the constructor \n hasReachedQuorum and isFull are now methods that can be called that will calculate the values of the desired variables \n boy I hope these newlines are actually newlines.

Chris-Pyles committed:
added function descriptor to matchGirls()

Chris-Pyles committed:
Added Session class

+ condensed Workshop class by moving some methods to the Session class
+ added totalRemainingCapacity method to Workshop class, returns a variable that is the remaining capacity of the workshop
- calcPopularity is no longer a method called by the constructor, as it would need to be declared before the constructor

Chris-Pyles committed:
Added Student Class

+ added Student class and its methods
    + assignWorkshop adds a workshop to a student's "Assigned" list
    + swapWorkshops switches the session times of two assigned workshops
    + numberAssigned is the number of workshops the student is assigned to
    + fullyAssigned is a boolean of whether or not the student has been assigned to 3 workshops
+ added subtractStudent method to Session class

Chris-Pyles committed:
Fixed magic numbers

changed "3" to "SESSIONS_PER_WORKSHOP" on line 257

Chris-Pyles committed:
Added makeStudentArray function

STUDENT_ARRAY variable is now constructed based on makeStudentArray

Chris-Pyles committed:
Fixed bug

changed "j" to "i" on line 198 to fix undefined error

Chris-Pyles committed:
Added init constructor to each class

+ added init constructor to each class that is called at the end of the class declaration
* Workshop.popularityScore is again calculated by a method that is now called by init
* Student.preferences is now calculated by a method that is called by init

Chris-Pyles committed:
Removed TODOs

beszel reviewed (approved):
Yes, good and cool
